### PR TITLE
Store SGE job IDs and Singleton DataAccess

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: src/main/scala/cromwell/engine/backend/jes/Run.scala [hunk 1]
```diff
@@ -75,7 +74,7 @@ object Run  {
 
       // Update the database state:
       val newBackendInfo = JesCallBackendInfo(db.CallStatus(ExecutionStatus.Running.toString), Option(JesId(run.jesId)), Option(JesStatus(currentStatus.toString)))
-      dataAccess.updateExecutionBackendInfo(run.workflowId, run.call, newBackendInfo)
+      DataAccess.instance.updateExecutionBackendInfo(run.workflowId, run.call, newBackendInfo)
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Maybe the following would be a bit less noisy as clients could just do `DataAccess().updateExecutionBackendInfo(...)`?

```scala
object DataAccess {
  def apply(): DataAccess = instance

  lazy val instance = new slick.SlickDataAccess()
}
```

File: src/main/scala/cromwell/engine/backend/jes/JesBackend.scala [hunk 2]
```diff
@@ -318,7 +317,8 @@ class JesBackend extends Backend with LazyLogging {
     }
   }
 
-  override def handleCallRestarts(restartableWorkflows: Seq[RestartableWorkflow], dataAccess: DataAccess)(implicit ec: ExecutionContext): Future[Any] = Future("FIXME")
+  // FIXME: Still needs FIXing
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/JesBackend.scala]: In the `Future`

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala [hunk 3]
```diff
@@ -42,14 +42,21 @@ class SgeBackend extends Backend with SharedFileSystem with LazyLogging {
...
+            updateSgeJobTable(backendCall, "Running", None, Option(sgeJobId))
+            pollForSgeJobCompletionThenPostProcess(backendCall, sgeJobId)
         }
       case Failure(ex) => FailedExecution(ex)
     }
   }
 
+  private def updateSgeJobTable(call: BackendCall, status: String, rc: Option[Int], sgeJobId: Option[Int]) = {
+    val backendInfo = SgeCallBackendInfo(CallStatus(status, rc), sgeJobId)
+    DataAccess.instance.updateExecutionBackendInfo(call.workflowDescriptor.id, call.call, backendInfo)
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: This seems to be just spinning off an asynchronous `Future` without considering the possibility of `Failure`.

File: src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala [hunk 4]
```diff
@@ -140,7 +140,7 @@ class SlickDataAccess(databaseConfig: Config, val dataAccess: DataAccessComponen
     case o => wdlType.fromWdlString(dbValue)
   }
 
-  override def shutdown() = database.shutdown
+  override def shutdown() = Future.successful(()) //database.shutdown
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala]: Why is this doing nothing useful?

File: src/test/scala/cromwell/CromwellTestkitSpec.scala [hunk 5]
```diff
@@ -63,11 +63,9 @@ with DefaultTimeout with ImplicitSender with WordSpecLike with Matchers with Bef
     }
   }
 
-  val dataAccess = DataAccess()
-
   override protected def afterAll() = {
     super.afterAll()
-    dataAccess.shutdown()
+    //dataAccess.shutdown()
```
Comment (mcovarr) [inline src/test/scala/cromwell/CromwellTestkitSpec.scala]: This should probably be deleted altogether

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala (hunk 3 above)
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: I guess I should at least compose the two DB updates so they can't happen in the wrong order... But otherwise what can you do if the status update fails? I don't want to fail the workflow...

File: src/main/scala/cromwell/engine/backend/jes/Run.scala (hunk 1 above)
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Personally I preferred dataAccess.instance.blahBlahBlah to DataAccess().blahBlahBlah since the former is clear that we're using a singleton instance whereas the latter looks like its creating a new one each time.

**Comment by cjllanwarne** [comment]
@mcovarr @kshakir I'm not sure which (if any) of the 'dataAccess.shutdown's should be left in? No tests work if any are left in but I'm fairly sure we want to close the singleton connection eventually...?

**Comment by mcovarr** [comment]
I wasn't concerned about the tests, just the production app.  If there's any flushing as part of `shutdown` I'd like to try to make that happen.

File: src/main/scala/cromwell/engine/backend/jes/Run.scala (hunk 1 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Yes, I think you're right.  I'm trying to find some implicity magic to support `DataAccess.updateExecutionBackendInfo` without having to tediously make the `DataAccess` companion object a delegate by hand.

Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: I also considered that... until I saw how many methods there were in it

File: src/main/scala/cromwell/server/WorkflowManagerSystem.scala [hunk 6]
```diff
@@ -12,15 +14,15 @@ trait WorkflowManagerSystem {
   implicit val actorSystem = ActorSystem(systemName)
 
   // For now there's only one WorkflowManagerActor so no need to dynamically name it
-  lazy val workflowManagerActor = actorSystem.actorOf(WorkflowManagerActor.props(dataAccess, backend), "WorkflowManagerActor")
+  lazy val workflowManagerActor = actorSystem.actorOf(WorkflowManagerActor.props(backend), "WorkflowManagerActor")
 
   // Lazily created as the primary consumer is the workflowManagerActor.
-  private lazy val dataAccess: DataAccess = DataAccess()
+  private lazy val dataAccess: DataAccess = DataAccess.instance
```
Comment (kshakir) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: I don't think this is used (private), and should be deleted?

File: src/main/scala/cromwell/server/WorkflowManagerSystem.scala [hunk 7]
```diff
@@ -12,15 +14,15 @@ trait WorkflowManagerSystem {
...
 
   /**
    * Should be called after the system is no longer in use.
    *
    * @return Non-blocking future that will eventually shut down this instance or return an error.
    */
-  def shutdown() = dataAccess.shutdown()
+  def shutdown() = Future.successful(()) //dataAccess.shutdown()
```
Comment (kshakir) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: Also not used.

File: src/test/scala/cromwell/webservice/CromwellApiServiceIntegrationSpec.scala [hunk 8]
```diff
@@ -16,13 +16,13 @@ import scala.concurrent.duration.Duration
...
-  val dataAccess = DataAccess()
-  val workflowManager = system.actorOf(WorkflowManagerActor.props(dataAccess, CromwellSpec.BackendInstance))
+  val dataAccess = DataAccess.instance
+  val workflowManager = system.actorOf(WorkflowManagerActor.props(CromwellSpec.BackendInstance))
   val version = "v1"
 
   override protected def afterAll() {
     super.afterAll()
-    Await.result(dataAccess.shutdown(), Duration.Inf)
+    //Await.result(dataAccess.shutdown(), Duration.Inf)
```
Comment (kshakir) [inline src/test/scala/cromwell/webservice/CromwellApiServiceIntegrationSpec.scala]: Delete this entire override. Git's revision control will save it for posterity if we need it again.

File: src/test/scala/cromwell/engine/db/slick/SlickDataAccessSpec.scala [hunk 9]
```diff
@@ -636,10 +635,5 @@ class SlickDataAccessSpec extends FlatSpec with Matchers with ScalaFutures {
         _ <- dataAccess.updateExecutionBackendInfo(workflowId, call, null)
       } yield ()).failed.futureValue should be(an[IllegalArgumentException])
     }
-
-    it should "shutdown the database" in {
-      assume(canConnect || testRequired)
-      dataAccess.shutdown().futureValue
```
Comment (kshakir) [inline src/test/scala/cromwell/engine/db/slick/SlickDataAccessSpec.scala]: Trace back this `dataAccess` variable (command click in IntelliJ). It **should** be shutdown, as SDAS is creating and testing the different connection pools.

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala [hunk 10]
```diff
@@ -131,25 +156,27 @@ class SgeBackend extends Backend with SharedFileSystem with LazyLogging {
...
     val abortFunction = killSgeJob(backendCall, sgeJobId)
     val waitUntilCompleteFunction = waitUntilComplete(backendCall)
     backendCall.callAbortRegistrationFunction.register(AbortFunction(abortFunction))
     val jobRc = waitUntilCompleteFunction
     logger.info(s"$tag SGE job completed (rc=$jobRc)")
-    (jobRc, backendCall.stderr.toFile.length) match {
-      case (r, _) if r == 143 => AbortedExecution // Special case to check for SIGTERM exit code - implying abort
+    val executionResult = (jobRc, backendCall.stderr.toFile.length) match {
+      case (r, _) if r == 143 =>
+        AbortedExecution // Special case to check for SIGTERM exit code - implying abort
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: why did you move this down?

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala [hunk 11]
```diff
@@ -42,14 +42,39 @@ class SgeBackend extends Backend with SharedFileSystem with LazyLogging {
...
           case (_, None) => FailedExecution(new Throwable(s"Could not parse Job ID from qsub output"))
-          case (_, Some(sgeJobId)) => pollForSgeJobCompletionThenPostProcess(backendCall, sgeJobId)
+          case (_, Some(sgeJobId)) =>
+            val updateDatabaseWithRunningInfo = updateSgeJobTable(backendCall, "Running", None, Option(sgeJobId))
+            val (executionResult, jobRc) = pollForSgeJobCompletionThenPostProcess(backendCall, sgeJobId)
+            updateDatabaseWithRunningInfo onComplete {
+              case _ =>
+                val completionStatus = statusString(executionResult)
+                val updateDatabaseWithCompletionInfo = updateSgeJobTable(backendCall, completionStatus, Option(jobRc), Option(sgeJobId))
+                updateDatabaseWithCompletionInfo onFailure recordDatabaseFailure(backendCall.call, completionStatus, jobRc)
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: I don't think you want to do all this stuff if this is a `Failure(e)`

File: src/main/scala/cromwell/engine/backend/jes/Run.scala (hunk 1 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Oh I didn't want to extend the trait, I was looking for a way to accomplish this with implicit magic.  And basically I failed, so never mind.  :frowning:

File: src/main/scala/cromwell/server/WorkflowManagerSystem.scala (hunk 7 above)
Comment (mcovarr) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: The production Cromwell code doesn't attempt to shutdown its `DataAccess` when it's being shut down, is that a problem?

Comment (geoffjentry) [inline src/main/scala/cromwell/server/WorkflowManagerSystem.scala]: We probably should. There's a ticket deep in the backlog to have some sort of proper shutdown functionality that latches in on the process' shutdown hook but it hasn't been a priority

File: src/main/scala/cromwell/engine/backend/jes/Run.scala (hunk 1 above)
Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: You could do `import DataAccess._` and then use `instance` directly ? If you call it `dataAccess` instead of `instance` it's almost seamless compared to how it was before.

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Yeah I like that.

File: src/main/scala/cromwell/engine/backend/jes/Run.scala [hunk 12]
```diff
@@ -8,6 +8,7 @@ import cromwell.engine.backend.jes.Run.{Failed, Running, Success, _}
 import cromwell.engine.db.{JesStatus, JesId, JesCallBackendInfo, DataAccess}
 import cromwell.util.google.GoogleScopes
 import org.slf4j.LoggerFactory
+import DataAccess._
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: If you rename the member in the DataAccess companion object to `dataAccess`, or  

```scala
import DataAccess.{slickDataAccess => dataAccess}
```

that would minimize the changes, and also make clear you're coding to the `DataAccess` trait and nothing particular about `SlickDataAccess`.  I'd suggest the former, since currently the member is named `slickDataAccess` but typed `DataAccess`.

Comment (kshakir) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Hm. I'm also a fan of _not_ minimizing the changes, but updating the variable name with something like `globalDataAccess`. Anything to indicate that the variable in question is actually some sort of imported singleton.

This feeling is based on emotional baggage after feeling all sorts of "wtf is going on"/"where did this come from" pain from Slick imports of variables+implicits, via things like [`import driver.api._`](https://github.com/broadinstitute/cromwell/blob/0d87d1e7ca0f64d82a630144b16da1d59dae811b/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala#L102).

Happy to defer to others though, as naming is yet another one of my nemeses.

File: src/test/scala/cromwell/engine/db/slick/SlickDataAccessSpec.scala [hunk 13]
```diff
@@ -11,6 +11,8 @@ import cromwell.engine._
 import cromwell.engine.backend.Backend.RestartableWorkflow
 import cromwell.engine.backend.local.{LocalBackend, LocalBackendCall}
 import cromwell.engine.backend.{Backend, ExecutionResult, StdoutStderr}
+import cromwell.engine.db.{CallStatus, ExecutionDatabaseKey, LocalCallBackendInfo}
+import cromwell.engine.workflow.CallKey
 import cromwell.engine.db.{CallStatus, DataAccess, ExecutionDatabaseKey, LocalCallBackendInfo}
 import cromwell.engine.workflow.{WorkflowOptions, CallKey}
```
Comment (kshakir) [inline src/test/scala/cromwell/engine/db/slick/SlickDataAccessSpec.scala]: Duplicated import statements above. ⌘⌥O to optimize the imports.

File: src/main/scala/cromwell/engine/backend/jes/Run.scala (hunk 12 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Okay, I was perhaps improperly conflating my objections here.  I don't like the name `slickDataAccess` when the type is `DataAccess`, as this does not reference anything special about the `SlickDataAccess` type.  I defer to your experience with overly magical implicit pain, let's do what we can to avoid that.  :smile:  `globalDataAccess` sounds fine to me.

Comment (Horneth) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: It's somewhat ambiguous I think because you indeed only need to depend on the `DataAccess` trait but when you import `DataAccess._` you actually make a dependency to Slick since the implementation of the trait is a `new SlickDataAccess()` in the `DataAccess` object.
Even if in most cases I think it overcomplexifies everything, IMO this is a case where the "cake pattern" method is pretty cool when you can say 
```scala
class Run { self: DataAccessDependency =>
...
}
```
with 
```scala
trait DataAccessDependency {
  val dataAccess: DataAccess // DataAccess trait
}
```
and then mix anything you want upon instantiation
```scala
trait SlickDataAccess extends DataAccessDependency {
  val dataAccess = new SlickDataAccess()
}
new Run(...) with SlickDataAccess
```
This makes it clear in the `Run` class that you only want "some kind of implementation for DataAccess", and let you choose at instantiation time which one you want to use.

Anyway that's probably out of scope for this ticket, changing to `globalDataAccess` should do it.

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala (hunk 11 above)
Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: Added a comment on line 51. Not sure why github hasn't pick up on the changes

**Comment by kshakir** [comment]
:+1:

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala (hunk 11 above)
Comment (kshakir) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: I believe the conversations tab shows the original context of the code. Over on the files tab, this comment thread has your comment on line 51. Will let @mcovarr comment further on if his issue was addressed.

**Comment by mcovarr** [comment]
:+1: assuming Travis is satisfied, delta any coverage complaints

File: src/main/scala/cromwell/engine/backend/jes/Run.scala [hunk 14]
```diff
@@ -75,7 +75,7 @@ object Run  {
 
       // Update the database state:
       val newBackendInfo = JesCallBackendInfo(db.CallStatus(ExecutionStatus.Running.toString), Option(JesId(run.jesId)), Option(JesStatus(currentStatus.toString)))
-      dataAccess.updateExecutionBackendInfo(run.workflowId, run.call, newBackendInfo)
+      globalDataAccess.updateExecutionBackendInfo(run.workflowId, run.call, newBackendInfo)
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Please see DSDEEPB-1462, this is writing backend info without taking index into account.

Comment (scottfrazer) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Are we handling that in this PR or separately?

Comment (cjllanwarne) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: If this is a separate ticket already, can I just merge this one?

Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/jes/Run.scala]: Let's address this in a separate PR and merge as is, but 0.12 shouldn't go out with the backends doing these index-unaware updates.  We could either just comment out these backend info updates (which wouldn't be any more broken than 0.11) or fix the problem and push back the release a bit.
